### ISSUE TITLE
[occm] should enforce all nodes in pool have the corresponding security group on node change

### ISSUE THREAD

**Issue by zifeo** [kubernetes/cloud-provider-openstack#2058]
**Is this a BUG REPORT or FEATURE REQUEST?**:

/kind bug

**What happened**:
I replaced one instance in my cluster and the load balancer pool correctly updated itself. However the security group was not applied to the newly created instance.

**What you expected to happen**:
Security groups should be enforced on any new node joining.

**How to reproduce it**:
1. create a load balancer
3. add a new node included in the pool
4. the matching security group was missing on the new

**Anything else we need to know?**:
`manage-security-groups: true` is in the config and the multiple Kubernetes load balancer share the same OpenStack load balancer using `loadbalancer.openstack.org/load-balancer-id`. 

**Environment**:
- openstack-cloud-controller-manager(or other related binary) version: helm chart 1.4.0
- OpenStack version: no access to version

**Comment by zifeo** [comment]
Relevant: https://github.com/kubernetes/cloud-provider-openstack/issues/490 @lingxiankong How did you manage those cases?

**Comment by jichenjc** [comment]
@zifeo unfortunately @lingxiankong not working on those area now, not sure he can reply to you on time

>>>the matching security group was missing on the new

if you see an issue ,could you please help provide PR if it's ok to you? seems this is a pretty reasonable case
the new nodes joined the group should get the security group anyway if it's not by bug

**Comment by zifeo** [comment]
@jichenjc Thanks for getting back. I likely do not have enough Go and OpenStack knowledge to be able to make this change in a reasonable amount of time sadly.

**Comment by jichenjc** [comment]
ok, then let me check what I can do , if I can find some area to improve, will submit a PR, thanks

**Comment by jichenjc** [comment]
@zifeo just to confirm, you use Octavia , not neutron -lbaas, correct? 
1.4 helm chart indicate latest chart so I think it's Octavia, just double confirm

**Comment by zifeo** [comment]
@jichenjc correct! `use-octavia: true`

**Comment by jichenjc** [comment]
ok, as I mentioned, original dev not work here any more so I am not very expert this area, but let me give a try
seems we do have a place to update the LB sec group
https://github.com/kubernetes/cloud-provider-openstack/blob/master/pkg/openstack/loadbalancer.go#L737

but in your case seems it's not done on time ,can you help
- enable --v = 5 in the OCCM daemonest to provide more logs 
- also, if rollout the daemonset will solve the problem (it should re-do the jobs to sync and maybe the new node will be updated)

**Comment by mdbooth** [comment]
/remove-lifecycle stale

@dulek Do you still want this one?

**Comment by dulek** [comment]
Yes! I need this to complete support for `manage-security-groups`.

/assign @dulek

**Comment by dulek** [comment]
Alright, so I've just tested this and in _most basic_ case this works. A new node is added to the LB pool, so `ensureOctaviaLoadBalancer()` is called and SGs are applied correctly.

The problem might be visible only with shared LBs as reporter wrote:
> the multiple Kubernetes load balancer share the same OpenStack load balancer using loadbalancer.openstack.org/load-balancer-id

But frankly I just tested this and it works. Either this got fixed by some other patch or there are some other circumstances when this happens. I'm closing this then:

/close
